Introduce time-masking to recurrent layers.

wxs opened the pull request:
This follows our discussion over in #176 where it was concluded that having a special "sentinel" value to represent something being masked out would be preferable to the approach I took in my earlier PR #239 

Comments welcome! I've not yet implemented this for any of the other recurrent layer types, pending some review.

<s>A note: my tests appear to take somewhat longer on my GPU machine than on my normal CPU-only machine. They're quite different, so this could be due to some other factor, but I thought it was worth mentioning in case I'm inadvertently using an operation that's unsafe for the GPU.</s> **Edit:** This was just because of the default `batch_size`, so I don't think is an issue

fchollet commented:
This look good so far! 

I don't immediately get the purpose of this:
```python
mask = T.addbroadcast(mask[:,:,np.newaxis], 2)
```

> A note: my tests appear to take somewhat longer on my GPU machine than on my normal CPU-only machine. They're quite different, so this could be due to some other factor, but I thought it was worth mentioning in case I'm inadvertently using an operation that's unsafe for the GPU.

Any way this hypothesis can be tested?

Also: please use spaces after commas and around arithmetic signs for code readability.

wxs commented:
mask = T.addbroadcast(mask[:,:,np.newaxis], 2)

This is important because the mask is `(time, nb_samples)` and the data is `(time, nb_samples, n_dim)`. I need to add the broadcast for the multiply in `_step` to get them to the same shape.

> Any way this hypothesis can be tested?

I altered the batch size and it went away, I think it's just the cost of CPU->GPU transfer

> Also: please use spaces after commas and around arithmetic signs for code readability.

On it :)

wxs commented:
Things that would be left to do on this:

 - Support masking in more recurrent types
 - Documentation
 - change the padding function

Anything else?

fchollet commented:
> change the padding function

I actually updated the padding function to use pre-padding as default earlier today.

Otherwise, we need to figure out what a good default mask value would be. '0.' is more aesthetic, but can it be proven safe? And what are the potential risks of using a value with a high absolute value?

fchollet commented on keras/layers/recurrent.py:
```diff
@@ -52,13 +53,16 @@ def get_output(self, train):
 
         x = T.dot(X, self.W) + self.b
+
+        mask = T.neq(x,self.mask_val).sum(axis=2)>0 # (time, nb_samples) matrix with a 1 for every unmasked entry
```
```python
mask = T.neq(x, self.mask_val).sum(axis=2) > 0
```

fchollet commented:
Thinking of the default mask value: it's definitely dangerous to have a non-0 padding as default for the `Embedding`, since that layer can often be used in combination with layers that don't do masking, such as `Dense` and `Convolution2D`.

wxs commented:
Waait I just realized a fundamental problem with this approach. Now we have the `SimpleRNN` output 0 for every masked value, so the mask doesn't propagate down multiple layers.

I need to have it instead output the mask, but then I *also* need to check h_tm1 to see if it's masked. Hmm.

fchollet commented:
> but then I also need to check h_tm1 to see if it's masked. 

Why?

wxs commented:
Well if I change the output of each SimpleRNN to be [-999] when masked, then that propagates properly down to the next layer of the deep net.

However, when I get an *unmasked* value , say at `t=3` it will incorporate in its step the value from `t=2`, which, since it was masked, is `[-999]`, not 0.

I could pass in mask_tm1 as another variable to `step`, and then do

     mask_t * self.activation(x_t + mask_tm1*T.dot(h_tm1, u))

that would fix it, at the cost of a bit more memory.

fchollet commented:
I believe this series of cascading issues could be avoided by reverting to the initial plan of a hard-coded mask value of 0., non configurable. Pros/cons?

wxs commented:
@fchollet I think the con that it won't work with ReLU activations is pretty significant!

wxs commented:
Take a look at 7f445d9, I think it solves the issue I described.

fchollet commented:
Good job on the fix, but I'd be worried about the performance impact, and the difficulty of generalizing that approach to SimpleDeepRNN.

I'm not entirely convinced that ReLU will cause issues with a mask value of 0. Also, accidental all-zero feature vectors would merely be skipped by recurrent units, which is not much of an issue (but again, even with ReLUs I don't think these accidents would ever happen).

If we are really worried about it, we could have ReLUs output start at epsilon instead of zero.

wxs commented:
> the difficulty of generalizing that approach to SimpleDeepRNN.

Hmm, I don't see why this is that hard to generalize to SimpleDeepRNN.

> I'm not entirely convinced that ReLU will cause issues with a mask value of 0

Why not? Definitely for low dimensions it would be a problem. If I have a 4-d hidden vector than 1/16 times I will trigger this behaviour!

And skipping some inputs would be very bad: in part because right now we ignore h_tm1 if we're masked, so we would lose all state built up to that point, and in part because it ignores the effect of the biases.

The epsilon approach might work, would need to think about some more.

wxs commented:
> but I'd be worried about the performance impact

Is the performance impact going to be different for a 0. mask? I still need to explicitly check if it's all 0 because the value that gets sent in to `_step` is not `X` but rather something like `W*X + b` so if X is all 0, you end up with `b` passed in to `_step`, which is not the mask_val.

wxs commented:
> Thinking of the default mask value: it's definitely dangerous to have a non-0 padding as default for the Embedding, since that layer can often be used in combination with layers that don't do masking, such as Dense and Convolution2D

Yeah that's definitely a concern. Could make masking opt-in rather than opt-out, i.e. force you to explicitly say `zero_mask=True` on the `Embedding` constructor.

wxs commented:
(I've made masking opt-in on the Embedding layer now)

wxs commented:
OK some updates: I've introduced masking to `SimpleDeepRNN` to show that the same approach works. As part of this I realized a better design for passing the mask, using the `taps` of scan.

Incidentally this also fixes a bug where you couldn't set `depth=1` in the `SimpleDeepRNN`

fchollet commented:
Looking good! `taps` does seem like the right way to get access to multiple past values of the mask efficiently. Nice catch.

Making the masking in Embedding opt-in seems reasonable, but `zero_is_mask=False` strikes me as unelegant (I don't have a better suggestion right now).

wxs commented:
OK another change: I realized that it makes more sense to have _step *apply* a mask, but
then to set the masked entries to mask_value outside of step. This
should be more efficient, but more importantly should make
implementations easier to understand.

Another nice effect: an alternative masking scheme can be introduced
without changing _step at all. I think you'll agree that this makes the _step functions very simple.

wxs commented:
> Making the masking in Embedding opt-in seems reasonable, but zero_is_mask=False strikes me as unelegant (I don't have a better suggestion right now).

Is it the naming? Or the general approach? Naming things is definitely something I struggle with :) perhaps `mask_zeros` or `zero_mask`?

wxs commented:
By the way: I'm still investigating some problems, this works well on my artificial test examples, but is currently failing on my actual application if I set `zero_is_mask=True` and pass in partially masked data. I don't know what the issue is yet so don't merge!

wxs commented:
I think masking will also need to be supported by `TimeDistributedDense` layers, incase someone builds an architecture that stacks them atop RNN layers. Thoughts?

fchollet commented:
> I don't know what the issue is yet so don't merge!

For merging I'll wait for everything to be tested; also it would be preferable to add masking to all recurrent layers at once : )

> I think masking will also need to be supported by TimeDistributedDense layers, incase someone builds an architecture that stacks them atop RNN layers. Thoughts?

Absolutely, that would be necessary. Or, we can by-pass the issue by using 0 as the default masking value (and offsetting our ReLUs by epsilon). 

Still, I believe anything in Keras that processes sequences should support masking. Including `TimeDistributedDense` and `time_distributed_softmax`.

wxs commented:
> Absolutely, that would be necessary. Or, we can by-pass the issue by using 0 as the default masking value (and offsetting our ReLUs by epsilon).

This won't bypass the issue. `TimeDistributedDense` includes a bias term so passing in 0 entries does not give 0 entries at the output. Still need to explicitly modify it for masking.

wxs commented:
@fchollet I had a thought, maybe you have some input here: if we made the mask value NaN it would be easier to find bugs where you accidentally pass masked data into a layer that does not support it. I would have to somewhat modify my work because `NaN != NaN` so the spots where I check equality to the mask value would need to instead use isNaN.

However I don't know if there are other concerns with using NaN, perhaps performance or otherwise.

wxs commented:
OK I think I've covered all the recurrent types now, and have also fixed the issue I was seeing (it was because I hadn't supported masking in Dropout yet). I've also cleaned the implementation up a bit. Any more comments?

fchollet commented:
> if we made the mask value NaN it would be easier to find bugs where you accidentally pass masked data into a layer that does not support it. I would have to somewhat modify my work because NaN != NaN so the spots where I check equality to the mask value would need to instead use isNaN.

Did you measure any performance when using NaN as mask value? 
NaN would preclude the customization of the mask value (which would probably be a good thing).

> However I don't know if there are other concerns with using NaN, perhaps performance or otherwise.

The propagation of NaN to any layer that doesn't do masking.

Besides NaN, what would be a sensible / elegant option for the default mask value? -999 is far from elegant.

wxs commented:
I haven't checked performance on NaN. I imagine that performance would vary quite a bit across different hardware. I'm also nervous because I know GPUs sometimes are a bit loose in their implementation of IEEE floating point and NaN is exactly the kind of corner case where we might see strange behaviour.

Propagation to the non-masking layers seems like a feature, not a bug, though. It is incorrect to pass masked data to a layer that does not support masking.

I think a potential problem would be that it might hide problems where your fit wasn't converging. I've had experiences, especially with ReLUs, of things blowing up to NaN. With NaN as mask you might not notice this explosion. Just like how I think "0" is an unsafe mask value because there's a much higher probability of inadvertent 0s than anything else, perhaps NaN is a problem as well.

> Besides NaN, what would be a sensible / elegant option for the default mask value? -999 is far from elegant.

I agree that `-999` seems somewhat arbitrary. At least it "stands out", if you accidentally hooked up a masked input to a layer that did not support masking you would probably notice the -999 while debugging.

I like that it's outside the range 0-1 because it is not a valid output from one of the "classic" activations. You would never see a -999 come out of a tanh or a sigmoid.

Which, incidentally, reminds me that probably the activation layers should all support masking.

wxs commented on keras/layers/recurrent.py:
```diff
@@ -58,12 +61,14 @@ def get_output(self, train):
         outputs, updates = theano.scan(
             self._step, # this will be called with arguments (sequences[i], outputs[i-1], non_sequences[i])
-            sequences=x, # tensors to iterate over, inputs to _step
+            sequences=[x, dict(input=padded_mask, taps=[-1])], # tensors to iterate over, inputs to _step
```
Incidentally, this `taps=[-1]` is technically unnecessary since a single negative tap will iterate over the input in the same exact way as just passing in the input without taps (see http://deeplearning.net/software/theano/library/scan.html and search for "abuse of notation"). But it maybe makes things more clear to explicitly name the taps anyway?

wxs commented:
This is now all working for me, and I'm using it on real data successfully, by the way.

fchollet commented on keras/layers/embeddings.py:
```diff
@@ -15,7 +17,7 @@ class Embedding(Layer):
         @out_dim: size of dense representation
     '''
-    def __init__(self, input_dim, output_dim, init='uniform', weights=None, W_regularizer=None, W_constraint=None):
+    def __init__(self, input_dim, output_dim, init='uniform', weights=None, W_regularizer=None, W_constraint=None, zero_is_mask=False, mask_val=default_mask_val):
```
This new parameter would also need to be included in get_config.

wxs replied:
OK I've done this across the layer types that take a mask_val.

fchollet replied:
Great! Except for `zero_is_mask` and `mask_val` in Embedding, apparently.

fchollet commented:
Two considerations:
- would masking cause downstream issues with any other layer type that we haven't yet considered? 
- could rounding issues, for instance on GPU, cause problems with computing a mask? i.e. could masked values deviate from the expected value by epsilon?

fchollet commented:
> This is now all working for me, and I'm using it on real data successfully, by the way.

What performance improvements do have see with masking compared to simple zero-padding with RNNs? Is computation time identical?

wxs commented:
> What performance improvements do have see with masking compared to simple zero-padding with RNNs? Is computation time identical?

I'm not seeing a noticeable run-time difference on the GPU, which makes sense because I think it still needs to run as many iterations of `scan` with and without masking. The main difference (and the reason I'm building this feature) is that my trained models are now insensitive to sequence length, which I need for my application so that I can keep feeding next words to the RNN and continue to get meaningful results.

wxs commented:
> could rounding issues, for instance on GPU, cause problems with computing a mask? i.e. could masked values deviate from the expected value by epsilon?

Yeah I'm definitely worried about this one. I think not, but let's think it through: `mask` is an int8 array of either 0 or 1. mask_val is a float32 on the GPU. If:

  - 1*mask_val is guaranteed to == mask_val
  - 0*mask_val is guaranteed to == 0.0
  - 0.0 + 1*mask_val is guaranteed to == mask_val
  - 1 - 1 is guaranteed to == 0.0

then I think we're good, because the mask is always applied thus `mask * normal_value + (1 - mask) * mask_val`. And I believe all those assumptions are safe.

Another question that I don't know the answer to: cheaper GPUs sometimes don't have ECC. If a bit were flipped in a mask maybe we'd have trouble? Of course I doubt the rest of the system is robust to that.

wxs commented:
> would masking cause downstream issues with any other layer type that we haven't yet considered

Yes. That was my main motivation to consider NaN so at least you might notice. I, for instance, had an issue where I had included a dropout layer, but had not supported masking in dropout.

The other option could be to have a flag that layers that support masking need to set, and then when you build the model throw an error if you try to hook up a masked layer to an unsupported layer.

fchollet commented:
> The other option could be to have a flag that layers that support masking need to set, and then when you build the model throw an error if you try to hook up a masked layer to an unsupported layer.

That sounds good. We should at least throw a warning. 

> Another question that I don't know the answer to: cheaper GPUs sometimes don't have ECC. If a bit were flipped in a mask maybe we'd have trouble? Of course I doubt the rest of the system is robust to that.

Right. That's also a reason why 0 (or possibly -epsilon) might be a better masking value, combined with strict enforcement of downstream masking support as per above, since in that case a random error would only have a tiny impact.

wxs commented:
> Right. That's also a reason why 0 (or possibly -epsilon) might be a better masking value, combined with strict enforcement of downstream masking support as per above, since in that case a random error would only have a tiny impact.

Is this true? Wouldn't flipping, for instance the MSB on a 0 also make a huge difference? Plus with biases I think it's dangerous to imagine that 0 is a neutral value anyway.

wxs commented:
> That sounds good. We should at least throw a warning.

Ok I'll look into that tomorrow.

kgzy commented:
@wxs when using TimeDistributeDense for multi-labelling tasks, will the loss function calculate the labels in the mask?

model.add(TimeDistributeDense(256,3,activation='tanh'))
model.add(Activation('time_distributed_softmax'))

when the length of a sequence is 5, then there are 5 labels, if the  max length is 10 , the y must be 10 dim and 5 of which is 0. will the time mask set the  the predict labels of the words in the mask to be 0?

wxs commented:
> when the length of a sequence is 5, then there are 5 labels, if the max length is 10 , the y must be 10 dim and 5 of which is 0. will the time mask set the the predict labels of the words in the mask to be 0?

@kgzy Hmm you're definitely right that the objectives should be able to handle masking. For my application I'm not using `return_sequences` on the final layer, so I don't need to handle the mask but for people that are evaluating an objective on the entire sequence the mask will be important.

There's a separate interesting question of *label masking*, i.e. what if for a sequence of length 10 you only have 4 labels you care about, but every input is valid? This PR doesn't cover that case but I think that would also be an interesting thing for someone to solve.

wxs commented:
OK I'm realizing that masking the *output* is closely related to PR #243. Here's what I propose:

In the case of multiple outputs from an RNN we have an output and label array of shapes (nb_samples, nb_timesteps, nb_dimensions). I propose reshaping that to (nb_samples * nb_timesteps, nb_dimensions), so each row consists of a sample at a point in time, and then masking thus:

```
reshaped_train = y_train.reshape((-1, y_train.shape[-1]))
masked_train = reshaped_train[y_mask.reshape((-1,1))]
# Do the same thing for self.y, and then
train_loss = self.loss(masked_pred, masked_y)
```

I think that if at this point I also had a weight matrix, I could multiply 
However, if `mask` is instead a weight matrix, rather than a boolean array, then this would also allow us to introduce weighting as per #243. However: I think that we have to then modify *every loss function*. Right now loss functions all compute their mean, and return a scalar but that doesn't work because we want a *weighted* mean across samples. I believe we want to instead have the loss functions return a (nb_samples * nb_timesteps, 1) array of the loss for each row, and then calculate the weighted mean on that, i.e.

    train_loss = (self.loss(self.y, self.y_train) * self.w).mean()

I can do this. Any concerns before I do?

fchollet commented:
> In the case of multiple outputs from an RNN we have an output and label array of shapes (nb_samples, nb_timesteps, nb_dimensions). I propose reshaping that to (nb_samples * nb_timesteps, nb_dimensions), so each row consists of a sample at a point in time

That seems highly confusing. It would be better imho to reshape the weight matrix. 

> I believe we want to instead have the loss functions return a (nb_samples * nb_timesteps, 1) array of the loss for each row, and then calculate the weighted mean on that

We could do that step inside the loss function. For consistency we definitely need to keep having the loss functions returns a scalar.

This is all introducing a lot of unforeseen complexity. I think there must be a better way. I will look into it...

wxs commented:
> We could do that step inside the loss function. For consistency we definitely need to keep having the loss functions returns a scalar.

Then you need to modify each loss function to take a weight/mask which also involves breaking changes to everything. What I'm currently doing is actually this:

    def get(identifier):
        obj_fn = get_from_module(identifier, globals(), 'objective')
        def weighted_obj_fun(y_true, y_pred, weights):
            # it's important that 0 * Inf == 0, not NaN, so I need to mask first
            masked_y_true = y_true[weights.nonero()]
            masked_y_pred = y_pred[weights.nonzero()]
            masked_weights = weights[weights.nonzero()]
            return (masked_weights * obj_fn(masked_y_true, masked_y_pred)).mean()

fchollet commented:
> What I'm currently doing is actually this:

`get(identifier)` is a shortcut, it should not include any logic. Any logic should go inside the loss functions. Modifying every loss function might not necessarily be a big issue.

At this point, I suggest we pause any implementation effort and take a step back to look at the global architecture implications of masking and weighting. We'll learned a lot, I'm sure we now have the bagage to figure out the "right" way to handle this (the simplest way).

wxs commented:
> At this point, I suggest we pause any implementation effort and take a step back to look at the global architecture implications of masking and weighting

Fine with me: I am using this branch in production for my projects but am happy to discuss alternatives (I may not be able to devote as much time to it going forward).

Do you have a preferred location to discuss the "right" approach, or do you want to keep the discussion here?

fchollet commented:
> Fine with me: I am using this branch in production for my projects but am happy to discuss alternatives (I may not be able to devote as much time to it going forward).
> Do you have a preferred location to discuss the "right" approach, or do you want to keep the discussion here?

Here is fine, I'm sure we'll end up merging this PR, maybe with slight modifications. I just want to make sure that we adequately weigh all design choices made so far.

wxs commented:
OK, here's a question: does it not arguably make *more* sense for loss functions to be a "per-datapoint" loss and have the mean calculated externally? Are there any examples of loss functions that are not averaged across all datapoints? (keep in mind that if such a loss existed you would not be able to calculate it by averaging across minibatches as we do now, external to the loss function).

wxs commented:
One alternative design. Layers have a get_mask() method (similar to get_output()) which defaults to returning  `None`. If you hook up a layer with a non-`None` `get_mask()` to one with a `None` `get_mask()` then it throws an error.

This fixes any issues around special sentinel values. Every mask-supported layer needs to be rewritten but that needs to happen anyway.

This design would fit nicely in with this PR, as it would be almost unchanged. The only changes would be in the details of how `get_mask` works, and the part where I do `(1 - mask) * self.mask_val`.

This does not address the objective weighting issue, which I think I should still solve closer to what I outlined above.

wxs commented:
My last couple commits show a fairly simple weighting scheme. If we like that approach, it is completely independent of the masking so I can put it up as a separate PR if we like it.

Note that if you're sending in masked sequences to an RNN for which `return_sequences=True` (as @kgzy is above) you would need to explicitly also send a weight matrix with 0s corresponding to the masked datapoints.

I could easily modify this to implicitly set weight=0 for all masked input if we like the approach more broadly, thereby simplifying the API when using a masked RNN input.

fchollet commented:
> One alternative design. Layers have a get_mask() method (similar to get_output()) which defaults to returning None. If you hook up a layer with a non-None get_mask() to one with a None get_mask() then it throws an error.

I like this idea, as it would allow for clean upward propagation of masks. The mask would be computed from the input values by the initial layer in a sequence, then would simply be passed on to the next layers. If we wanted to switch from a masking system *not* based on input values, that would be easy to achieve since we would just have to modify that initial step. 

So I think this is a great design idea.

We could avoid all code redundancy by introducing a MaskingLayer subclass that exposes the universal `get_mask()` method.

wxs commented:
> I don't think weighting should happen at validation / testing time. Loss weighting is a training-specific strategy...

I can see think of a few cases where you would want to weight on validation data.

1) it lets your weight matrix double as an *output* mask. Output masks are distinct from input masks (which the rest of this PR is about) because they include the case where only some of your outputs are labelled.

2) you can use weighting to normalize your class counts, in which case you would also want that weighting applied at validation time.

3) Perhaps weighting would be useful as part of a boosted architecture? (I'm less sure of this case)

wxs commented:
> We could avoid all code redundancy by introducing a MaskingLayer subclass that exposes the universal get_mask() method

Great, well I think I'll have some time to look into this tomorrow.

kgzy commented:
It's great!  The weight matrix makes the loss function more flexible! Thank you

kgzy commented:
@wxs there is a bug in model.py , function 'fit'.
   when given validation_split and weights is none , the variable weight_val is not defined, but used in line 230.

wxs commented:
Oops ok will add a testcase for that.

wxs commented:
OK well that was exhausting: got masking built under the new system :) the user-facing API is almost identical but masks are now sent differently.

I think this is much safer, I'm glad we moved away from sentinel values.

fchollet commented:
I don't understand why there is  `get_output_mask` as well as an `get_input_mask`. A single `get_mask` ought to be sufficient? Especially given that both default methods return the same thing and `get_input_mask` is basically never used.

```python
def get_output_mask(self, train=None):
    return self.previous.get_output_mask(train)

def get_input_mask(self, train=None):
    return self.previous.get_output_mask(train)
```

Also, the default methods call the `previous` attribute which may not exist. The way `get_input` works is by checking for `previous` and if there is no previous layer, defaulting to the `input` tensor. In the same way I believe we should check for `previous` here...

`mask_zero` in the `Embedding` is definitely a more elegant choice of name ;)

I'll dive deeper into this as soon as I get some time.

wxs commented:
> I don't understand why there is get_output_mask as well as an get_input_mask. A single get_mask ought to be sufficient

I had that originally. The problem is that when I actually *use* the mask I need the previous layer's mask, because it's my input that's coming in masked. Alternatively I could do:

        mask = self.previous.get_output_mask(train)

in `get_padded_shuffled_mask`. I just thought this might be more clear.

wxs commented:
> Also, the default methods call the previous attribute which may not exist. The way get_input works is by checking for previous and if there is no previous layer, defaulting to the input tensor. In the same way I believe we should check for previous here

Yeah we could check for previous and return `None` if it is not there.

fchollet commented:
For the record, do you consider this PR to be merge-ready right now? If yes, I will pull it and do an in-depth review, then merge.

wxs commented:
Yes, though I've not tested the latest version on the GPU yet. 

> On Jun 25, 2015, at 19:47, François Chollet <notifications@github.com> wrote:
> 
> For the record, do you consider this PR to be merge-ready right now? If yes, I will pull it and do an in-depth review, then merge.
> 
> —
> Reply to this email directly or view it on GitHub.
>

wxs commented:
OK I had to make one small change to get it working on the GPU again. Should be good to go now.

One annoying thing: my tests are not deterministic, so they sometimes fail just by accidentally getting stuck in a local minimum. Perhaps you have better ideas for testing patterns.

wxs commented:
One thing I did not do here, but which is worth discussing. For applications like that of @kgzy wherein `return_sequences=True` it might be nice to use the input mask as an output mask as well (basically multiply the weights matrix by the mask matrix in the last stage). Right now you would need to explicitly mask your labels as well.

fchollet commented:
Looking at it now. Attempted to run the `imdb_lstm.py` example (with no changes):
```
ValueError: Input dimension mis-match. (input[0].shape[1] = 16, input[1].shape[1] = 1)
Apply node that caused the error: Elemwise{Composite{(i0 * ((i1 * log(i2)) + (i3 * log(i4))))}}(InplaceDimShuffle{x,0}.0, AdvancedSubtensor1.0, Elemwise{clip,no_inplace}.0, Elemwise{sub,no_inplace}.0, Elemwise{sub,no_inplace}.0)
Toposort index: 421
Inputs types: [TensorType(float64, row), TensorType(float64, matrix), TensorType(float64, matrix), TensorType(float64, matrix), TensorType(float64, matrix)]
Inputs shapes: [(1, 16), (16, 1), (16, 1), (16, 1), (16, 1)]
Inputs strides: [(128, 8), (8, 8), (8, 8), (8, 8), (8, 8)]
Inputs values: ['not shown', 'not shown', 'not shown', 'not shown', 'not shown']
Outputs clients: [[Sum{acc_dtype=float64}(Elemwise{Composite{(i0 * ((i1 * log(i2)) + (i3 * log(i4))))}}.0)]]
```

Same dimension mismatch error when enabling masking (`mask_zero=True` in Embedding).

fchollet commented:
Have the changes been tested with recurrent layers set to `return_sequences=False`?

wxs commented:
> Have the changes been tested with recurrent layers set to return_sequences=False?

Yes for my production system `return_sequences=False`. Also my first test-case has `return_sequences=False`.

Though if you ever had a *completely* masked input your results wouldn't make sense I guess. `return_sequences=False` layers opt to not send their results to the output.

wxs commented:
> Looking at it now. Attempted to run the imdb_lstm.py example (with no changes):

Are you sure? I just ran it and it worked.

I'm on Python3. Could there be a subtle 2/3 difference?

fchollet commented:
Doesn't work for me on Python 3 either. I'm up to date with `wxs/keras.git mask_value`.

wxs commented:
Sorry you're right I was having a pythonpath issue. Looking into it.

wxs commented:
@fchollet ok I've fixed that: it's because binary_crossentropy and categorical_crossentropy were giving differently shaped results.

fchollet commented on keras/layers/recurrent.py:
```diff
@@ -38,19 +61,21 @@ def __init__(self, input_dim, output_dim,
             self.set_weights(weights)
 
-    def _step(self, x_t, h_tm1, u):
+    def _step(self, x_t, mask_tm1, h_tm1, u):
```
Shouldn't this variable be named `mask_t` everywhere? Or am I missing something?

wxs replied:
No, it represents the mask of the previous timestep. We use that mask value to decide whether to include h_tm1 in our calculation of h_t. If mask_tm1 is 0 then h_tm1 is invalid.

fchollet commented:
On the IMDB LSTM example, you would expect masking padded values to improve accuracy compared to incorporating the padding into the computations. At the very least, you wouldn't expect masking to perform worse than no masking.

In practice, it performs considerably worse (by about 10 percentage points in accuracy). This could point to implementation correctness issues with LSTM (and potentially other layers as well). 

What do you think?

wxs commented:
> In practice, it performs considerably worse (by about 10 percentage points in accuracy). This could point to implementation correctness issues with LSTM (and potentially other layers as well).

It could. It could also point to the importance of then initial state. If most of the imdb samples are pre-padded with many 0s, then in the non-masked case it can learn a representation of 0 that corresponds to a reasonable initial state.

For my use-case I define a new "start-of-sentence" token that I prepend to each sequence. This is a common practice. So every training example becomes PAD PAD PAD <S> this is a sequence. If you update the IMDB example to prepend a special token to each training example does the gap close? (I can't try right now).

fchollet commented:
Sure, let me check the effect of adding a start character.

wxs commented:
@fchollet Oh wait another issue: you're *also* using 0 to represent "out-of vocabulary" tokens here:

        X = [[0 if (w >= nb_words or w < skip_top) else w for w in x] for x in X]

You obviously need to use a different token there. I use 0 for PAD, 1 for OOV, and 2 for start-of-sentence.

fchollet commented:
Right. We'll definitely need to update the preprocessing utils and dataset utils to support different OOV chars and start of sequence chars.

I'm running new tests with both features right now.

fchollet commented:
Adding an OOV char and a start char almost completely closes the gap. Performance is still slightly degraded compared to no masking, by 1-2 percentage points (which is on the higher end of the error margin for the dataset). 

I will add support for OOV/Start chars everywhere and merge this a bit later today, I expect.

fchollet commented:
I take that back, it seems my tests had indeterministic factors that I didn't notice at first. Here is a 100% apples to apples comparison (exact same data, same initial weights, different characters for OOV, start, and padding):

without `mask_zero`:
```
Train on 18000 samples, validate on 2000 samples
Epoch 0
18000/18000 [==============================] - 369s - loss: 0.5842 - acc.: 0.6858 - val. loss: 0.4347 - val. acc.: 0.8105
Epoch 1
18000/18000 [==============================] - 379s - loss: 0.3408 - acc.: 0.8634 - val. loss: 0.4350 - val. acc.: 0.8210
Epoch 2
18000/18000 [==============================] - 387s - loss: 0.2252 - acc.: 0.9201 - val. loss: 0.4357 - val. acc.: 0.8210
```

with `mask_zero`:
```
Epoch 0
18000/18000 [==============================] - 371s - loss: 0.6510 - acc.: 0.6282 - val. loss: 0.6469 - val. acc.: 0.6830
Epoch 1
18000/18000 [==============================] - 381s - loss: 0.6423 - acc.: 0.6389 - val. loss: 0.6322 - val. acc.: 0.6515
Epoch 2
18000/18000 [==============================] - 382s - loss: 0.5527 - acc.: 0.7304 - val. loss: 0.5130 - val. acc.: 0.7630
```

I pushed all the changes I've made so far to https://github.com/fchollet/keras/tree/wxs-mask_value

Thoughts?

wxs commented:
Hmm. Well I don't think it makes sense to call this an "apples to apples" comparison simply because the matrix weights are the same, since the matrices have different meaning in the two cases.

I removed the seeding, and trained the model both with and without zero_mask 4 times respectively.

Without masking final val. accuracy
0.8125 0.8115 0.7965 0.8140 (mean: 0.8086)

WIth masking final val. accuracy
0.8135 0.8220 0.7825 0.7655 (mean: 0.7958) 

which seem pretty close to me.

I also realized that you truncate when padding differently than I do (I opened issue #287 about this, for discussion) which means you give a *small* amount more information to the unmasked classifier than to the masked one. In the unmasked classifier the presence of an initial 0 now means "this sequence was not truncated".

The masked classifier doesn't get to see that 0, so doesn't know if it's on a truncated or untruncated sequence. However if you truncated off the *left* when padding on the left, then the presence or absence of the start-of-sentence indicator would contain this same information.

I don't know if that actually explains the difference between the classifiers, but I'm not convinced that the difference between them is statistically significant anyway.

I'm going to run an experiment with the alternate padding, just to see if it makes a difference.

fchollet commented:
> I'm not convinced that the difference between them is statistically significant anyway.

I'll do randomized testing to check this, in this case.

Do you have any reproducible example where masking brings a significant performance improvement? I imagine that shouldn't be hard to come across.

wxs commented:
> Do you have any reproducible example where masking brings a significant performance improvement? I imagine that shouldn't be hard to come across.

Well the best example I have is that *without* masking you can end up training classifiers that are very sensitive to sequence length, because they've only ever been trained on sequences of exactly that length. This is problematic: say I'm doing a sentiment classifier that I want to have outputing predictions on live data. If my classifier is sensitive to sequence length, then as each token comes in I need to take a 20-token history slice and run 20 steps of the classifier to get a meaningful result.

If my classifier is *not* sensitive to sequence length, I can just feed in the latest token to the previous state and get another meaningful result, making my resultant classifier suitable for real-time data.

This is why I built the feature, actually. I was getting perfectly good classifiers, except they needed inputs of length exactly 20. Now I'm getting equally good classifiers that can take sequences of arbitrary length.

wxs commented:
An idea of how, once you have masking, you could extend to also improve training speeds:

 - Define a micro-batch size (sorry that's a bad name).
 - For each mini-batch, sort the data points by their sequence length
 - Now divide your mini-batch into micro-batches. datapoints in a micro-batch will be of similar length
 - truncate each micro-batch to its longest element
 - backprop your micro-batches, average the gradients across a mini-batch before taking a gradient step to get the same results as you would have had without this optimization (protects against weird biases to do with calculating gradient steps based on non-random data subsets)
 - Now, because of the truncation, the Theano scan doesn't need to make as many steps, on average.

I might implement this for my purposes, though I acknowledge that it introduces a kind of ugly/confusing API around micro-batches vs. mini-batches.

wxs commented:
I changed the truncation to truncate the beginning rather than the end, and got better results for every trial on both masked and unmasked datasets. Results for 4 trials below:

Unmasked: 0.8245 0.8265 0.8235 0.8230
Masked: 0.8090 0.8235 0.8225 0.8115

Unmasked still seems to have a slight edge, which is interesting. It could be a bug, or perhaps just an idiosyncrasy of this particular architecture? Will think about it a bit more.

fchollet commented:
Interesting. It would be great to have a couple more datapoints than just the IMDB example.

fchollet commented:
> If my classifier is not sensitive to sequence length, I can just feed in the latest token to the previous state and get another meaningful result, making my resultant classifier suitable for real-time data.

But isn't the memory of RNN layers being reset with each minibatch in the current setup? Or did you make changes to make the RNN layers stateful?

fchollet commented:
I ran further randomized testing and the performance with masking appears to be consistent with no masking (with just a slight dip --unclear what causes it). Besides, the implementation looks fine to me.

I also added masking support for the new recurrent layers, the JZSs.  The PR is likely getting merged later today.

fchollet commented:
Merged. I also used to occasion to re-introduce class weighting, thanks to your `weighted_objective` function.

Thank you for the valuable contribution.

wxs commented:
Great! Happy to contribute. Ping me if there are any issues with it you'd like me to take a look at.

vzhong commented:
@wxs can you please add some documentations for `MaskedLayer` as well as its API, see how this is now a rather significant interface?

wxs commented:
@vzhong Yes I would like to get to that, although it won't be in the next couple of days. Luckily the user-facing APIs for the unmasked case are still identical.

vzhong commented:
Thanks @wxs 

Would you mind taking a quick look at #320 ? I suspect it is related to this change. Perhaps you can spot an easy fix?
